File: leetsync/3434-find-the-number-of-distinct-colors-among-the-balls/find-the-number-of-distinct-colors-among-the-balls.go
func queryResults(limit int, queries [][]int) []int {
	n := len(queries)
	result := make([]int, 0, n)
	ballToColorMap := map[int]int{}
	colorToFreqMap := map[int]int{}
	for _, query := range queries {
		ball := query[0]
		color := query[1]
		if previousColor, ok := ballToColorMap[ball]; ok {
			colorToFreqMap[previousColor]--
			if colorToFreqMap[previousColor] == 0 {
				delete(colorToFreqMap, previousColor)
			}
		}
		ballToColorMap[ball] = color
		colorToFreqMap[color]++
		result = append(result, len(colorToFreqMap))
	}
	return result
}

File: leetsync/3434-find-the-number-of-distinct-colors-among-the-balls/find-the-number-of-distinct-colors-among-the-balls.java
class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        int n = queries.length;
        int[] result = new int[n];
        Map<Integer, Integer> ballToColorMap = new HashMap<Integer, Integer>();
        Map<Integer, Integer> colorToFreqMap = new HashMap<Integer, Integer>();
        for(int i=0; i<n; i++){
            int[] query = queries[i];
            int ball = query[0];
            int color = query[1];
            if(ballToColorMap.containsKey(ball)){
                int previousColor = ballToColorMap.get(ball); // already colored ball
                colorToFreqMap.put(previousColor, colorToFreqMap.get(previousColor)-1);
                if(colorToFreqMap.get(previousColor)==0){
                    colorToFreqMap.remove(previousColor);
                }
            }
            ballToColorMap.put(ball, color);
            colorToFreqMap.put(color, colorToFreqMap.getOrDefault(color, 0)+1);
            result[i] = colorToFreqMap.size();
        }
        return result;
    }
}